### PR TITLE
Support non-constant timezone arguments in toString() and formatDateTime()

### PR DESCRIPTION
**Pull request by yariks5s**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
SQL functions "toString(datetime)" and "formatDateTime()" now support non-constant timezone arguments.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit c75cf6e1541afaf9ec72204af60356cb113adc81 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53680/c75cf6e1541afaf9ec72204af60356cb113adc81/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by alexey-milovidov** [comment]
Does it work?

```
SELECT toString(now(), number % 2 ? 'America/Los_Angeles' : 'Europe/Amsterdam') FROM numbers(5);
SELECT formatDateTime(now(), '%H%i%S', number % 2 ? 'America/Los_Angeles' : 'Europe/Amsterdam') FROM numbers(5);
SELECT toHour(now(), number % 2 ? 'America/Los_Angeles' : 'Europe/Amsterdam') FROM numbers(5);
```

**Comment by Avogar** [comment]
Please, add tests. You can add them to `tests/queries/0_stateless/` directory. You can take this [test](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01592_toUnixTimestamp_Date.sql) as an example

File: src/Functions/FunctionsConversion.h [hunk 1]
```diff
@@ -1844,20 +1844,20 @@ class FunctionConvert : public IFunction
...
             || std::is_same_v<ToDataType, DataTypeDate32>
-            // toDateTime(value[, timezone: String])
+            // toDateTime(value[, timezone: String(const/non-const)])
             || std::is_same_v<ToDataType, DataTypeDateTime>
-            // toDateTime64(value, scale : Integer[, timezone: String])
+            // toDateTime64(value, scale : Integer[, timezone: String(const/non-const)])
             || std::is_same_v<ToDataType, DataTypeDateTime64>)
         {
-            optional_args.push_back({"timezone", &isString<IDataType>, &isColumnConst, "const String"});
+            optional_args.push_back({"timezone", &isString<IDataType>, nullptr, "const String/String"});
```
Comment (rschu1ze) [inline src/Functions/FunctionsConversion.h:1860]: `const String/String` sounds wrong. Typo?

File: src/Functions/FunctionsConversion.h [hunk 2]
```diff
@@ -1844,20 +1844,20 @@ class FunctionConvert : public IFunction
             mandatory_args.push_back({"scale", &isNativeInteger<IDataType>, &isColumnConst, "const Integer"});
         }
 
-        // toString(DateTime or DateTime64, [timezone: String])
+        // toString(DateTime or DateTime64, [timezone: String (const/non-const)])
```
Comment (rschu1ze) [inline src/Functions/FunctionsConversion.h:1847]: Could we remove "const/non-const" in l. 1847-1857? That's conceptually the default anyways. If needed, you can add a comment above l. 1860 that the timezone argument is deliberately non-const or const.

File: src/Functions/extractTimeZoneFromFunctionArguments.cpp [hunk 3]
```diff
@@ -30,11 +30,11 @@ std::string extractTimeZoneNameFromColumn(const IColumn * column, const String &
 }
 
 
-std::string extractTimeZoneNameFromFunctionArguments(const ColumnsWithTypeAndName & arguments, size_t time_zone_arg_num, size_t datetime_arg_num, bool allow_nonconst_timezone_arguments)
+std::string extractTimeZoneNameFromFunctionArguments(const ColumnsWithTypeAndName & arguments, size_t time_zone_arg_num, size_t datetime_arg_num, [[maybe_unused]] bool allow_nonconst_timezone_arguments)
```
Comment (rschu1ze) [inline src/Functions/extractTimeZoneFromFunctionArguments.cpp:33]: Argument `allow_nonconst_timezone_arguments` is no longer used. Could you maybe just remove it (and adjust the callers)?

File: src/Functions/extractTimeZoneFromFunctionArguments.cpp [hunk 4]
```diff
@@ -58,10 +58,24 @@ const DateLUTImpl & extractTimeZoneFromFunctionArguments(const ColumnsWithTypeAn
...
     if (arguments.size() == time_zone_arg_num + 1)
     {
-        std::string time_zone = extractTimeZoneNameFromColumn(arguments[time_zone_arg_num].column.get(), arguments[time_zone_arg_num].name);
-        if (time_zone.empty())
-            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Provided time zone must be non-empty and be a valid time zone");
-        return DateLUT::instance(time_zone);
+        [[maybe_unused]]const auto * left_const = typeid_cast<const ColumnConst *>(arguments[time_zone_arg_num].column.get());
+        if (left_const)
+        {
+            std::cerr << "=================" << arguments[time_zone_arg_num].column.get()->dumpStructure() << "==================" << std::endl;
```
Comment (rschu1ze) [inline src/Functions/extractTimeZoneFromFunctionArguments.cpp:64]: Spams the traces --> remove

File: src/Functions/extractTimeZoneFromFunctionArguments.cpp [hunk 5]
```diff
@@ -16,25 +16,22 @@ namespace ErrorCodes
     extern const int ILLEGAL_TYPE_OF_ARGUMENT;
 }
 
-
-std::string extractTimeZoneNameFromColumn(const IColumn * column, const String & column_name)
+std::string extractTimeZoneNameFromColumn(const IColumn * column, [[maybe_unused]] const String & column_name)
```
Comment (rschu1ze) [inline src/Functions/extractTimeZoneFromFunctionArguments.cpp:19]: Argument allow_nonconst_timezone_arguments is no longer used. Could you maybe just remove it (and adjust the callers)?

File: src/Functions/extractTimeZoneFromFunctionArguments.cpp [hunk 6]
```diff
@@ -16,25 +16,22 @@ namespace ErrorCodes
...
-                        "Illegal column {} of time zone argument of function, must be a constant string",
-                        column_name);
+        return "";
 
     return time_zone_column->getValue<String>();
 }
 
 
-std::string extractTimeZoneNameFromFunctionArguments(const ColumnsWithTypeAndName & arguments, size_t time_zone_arg_num, size_t datetime_arg_num, bool allow_nonconst_timezone_arguments)
+std::string extractTimeZoneNameFromFunctionArguments(const ColumnsWithTypeAndName & arguments, size_t time_zone_arg_num, size_t datetime_arg_num, [[maybe_unused]] bool allow_nonconst_timezone_arguments)
```
Comment (rschu1ze) [inline src/Functions/extractTimeZoneFromFunctionArguments.cpp:30]: Same here for `allow_nonconst_timezone_arguments`.

File: src/Functions/extractTimeZoneFromFunctionArguments.cpp [hunk 7]
```diff
@@ -16,25 +16,22 @@ namespace ErrorCodes
...
-std::string extractTimeZoneNameFromColumn(const IColumn * column, const String & column_name)
+std::string extractTimeZoneNameFromColumn(const IColumn * column, [[maybe_unused]] const String & column_name)
 {
     const ColumnConst * time_zone_column = checkAndGetColumnConst<ColumnString>(column);
 
     if (!time_zone_column)
-        throw Exception(ErrorCodes::ILLEGAL_COLUMN,
-                        "Illegal column {} of time zone argument of function, must be a constant string",
-                        column_name);
+        return "";
```
Comment (rschu1ze) [inline src/Functions/extractTimeZoneFromFunctionArguments.cpp:24]: Did you check that all callers can deal with "" as time zone? E.g. Functions/FunctionsCodingULID.cpp doesn't seem prepared. It throws an exception for empty time zone strings, not really what the user expects.

Comment (rschu1ze) [reply src/Functions/extractTimeZoneFromFunctionArguments.cpp:24]: Also suggest to carefully check all callers of `extractTimeZoneNameFromFunctionArguments()` (which you changed below)

File: src/Functions/IFunction.cpp [hunk 8]
```diff
@@ -321,7 +321,12 @@ ColumnPtr IExecutableFunction::execute(const ColumnsWithTypeAndName & arguments,
             const auto * column_sparse = checkAndGetColumn<ColumnSparse>(arguments[i].column.get());
             /// In rare case, when sparse column doesn't have default values,
             /// it's more convenient to convert it to full before execution of function.
-            if (column_sparse && column_sparse->getNumberOfDefaultRows())
+            const ColumnSparse * col_sparse = nullptr;
+            if (!column_sparse)
```
Comment (rschu1ze) [inline src/Functions/IFunction.cpp:325]: Not clear why this logic is necessary.

File: src/Functions/extractTimeZoneFromFunctionArguments.cpp (hunk 7 above)
Comment (yariks5s) [reply src/Functions/extractTimeZoneFromFunctionArguments.cpp]: Yes, you are right, I've just fixed this issue

**Comment by yariks5s** [comment]
There was a suggestion, which, I think, make sense
> The problem is (see the previous PRs) that the timezone is part of the data type and not part of the value. This can lead to confusion for users, so it is better to disallow non-const timezones as a general rule. One may argue that `formatDateTime` and `toString(dt, tz)` can be exceptions as they produces a string. Suggest to keep `toDateTime` as is - it produces a `DateTime`.

**Comment by alexey-milovidov** [comment]
@rschu1ze About the functions that produce DateTime from String - they can support non-constant timezone but return the DateTime type with the server's timezone. It will lead to some inconsistency, but it is still good. We can also postpone it and add in another pull request.

File: src/Functions/formatDateTime.cpp [hunk 9]
```diff
@@ -855,11 +855,18 @@ class FunctionFormatDateTimeImpl : public IFunction
...
     ColumnPtr executeType(const ColumnsWithTypeAndName & arguments, const DataTypePtr &) const
     {
-        auto * times = checkAndGetColumn<typename DataType::ColumnType>(arguments[0].column.get());
+        auto non_const_datetime = arguments[0].column->convertToFullColumnIfConst();
+        auto * times = checkAndGetColumn<typename DataType::ColumnType>(non_const_datetime.get());
         if (!times)
             return nullptr;
 
         const ColumnConst * format_column = checkAndGetColumnConst<ColumnString>(arguments[1].column.get());
+        const ColumnConst * time_zone_column = nullptr;
```
Comment (rschu1ze) [inline src/Functions/formatDateTime.cpp:864]: Please move l. 870 - 873 (the check for `~format_column`) up to l. 853.

File: src/Functions/formatDateTime.cpp [hunk 10]
```diff
@@ -855,11 +855,18 @@ class FunctionFormatDateTimeImpl : public IFunction
     template <typename DataType>
     ColumnPtr executeType(const ColumnsWithTypeAndName & arguments, const DataTypePtr &) const
     {
-        auto * times = checkAndGetColumn<typename DataType::ColumnType>(arguments[0].column.get());
+        auto non_const_datetime = arguments[0].column->convertToFullColumnIfConst();
```
Comment (rschu1ze) [inline src/Functions/formatDateTime.cpp:858]: Curious, why we now unpack const columns?

File: src/Functions/formatDateTime.cpp [hunk 11]
```diff
@@ -941,6 +952,13 @@ class FunctionFormatDateTimeImpl : public IFunction
         auto * pos = begin;
         for (size_t i = 0; i < vec.size(); ++i)
         {
+            if (!time_zone_column && arguments.size() > 2)
```
Comment (rschu1ze) [inline src/Functions/formatDateTime.cpp:955]: `time_zone_column` should really be called `const_time_zone_column`.

**Comment by rschu1ze** [comment]
I had #50834 \+ #48471 in mind, both PRs are for `toTimezone()` which is maximally confusing when used with non-const timezones. But yes, there may be more string-to-DT functions where a non-const timezone may make sense. Best decided on a case-by-case basis, in another PR and well documented.

File: src/Functions/formatDateTime.cpp [hunk 12]
```diff
@@ -941,6 +953,13 @@ class FunctionFormatDateTimeImpl : public IFunction
         auto * pos = begin;
         for (size_t i = 0; i < vec.size(); ++i)
         {
+            if (!const_time_zone_column && arguments.size() > 2)
+            {
+                if (!arguments[2].column.get()->getDataAt(i).toString().empty())
+                    time_zone = &DateLUT::instance(arguments[2].column.get()->getDataAt(i).toString());
+                else
+                    throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Provided time zone must be non-empty and be a valid time zone");
```
Comment (rschu1ze) [inline src/Functions/formatDateTime.cpp:961]: Minor: I know this message was borrowed from extractTimeZoneFunctionArgument.cpp, but the error text `[...] and be a valid time zone` is a bit weird. We are only checking (here and in the other file) that the time zone is non-empty. We are not checking if it is "valid", e.g. "Fantasia/DisneyLand" is technically not rejected here (it is rejected in l. 959. Let's say in the error message only "... must be non-empty." (also in the other file, please)

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
@yariks5s The [performance regressions](https://s3.amazonaws.com/clickhouse-test-reports/53680/7f4fa192c571079721a79c9de13cddaac6696e32/performance_comparison_[3_4]/report.html) in `toString()` look (on the surface) related. Could you take a look if they are connected to this PR?

File: src/Functions/FunctionsConversion.h [hunk 13]
```diff
@@ -1854,7 +1883,7 @@ class FunctionConvert : public IFunction
             // toDateTime64(value, scale : Integer[, timezone: String])
             || std::is_same_v<ToDataType, DataTypeDateTime64>)
         {
-            optional_args.push_back({"timezone", &isString<IDataType>, &isColumnConst, "const String"});
+            optional_args.push_back({"timezone", &isString<IDataType>, nullptr, "String"});
```
Comment (rschu1ze) [inline src/Functions/FunctionsConversion.h:1886]: Maybe I miss something but is this PR not supposed to make the timezone argument non-const for `toString()`, but not for date-time-like destination types like`toDate32()`, `toDateTime()`? Looking at 1875-1884, this isn't the case.

EDIT: The other cases seem covered by l. 1950ff, can you confirm?

**Comment by rschu1ze** [comment]
Still a performance regression :-(

File: src/Functions/FunctionsConversion.h (hunk 13 above)
Comment (yariks5s) [reply src/Functions/FunctionsConversion.h:1886]: Confirm, toString() case covered on line 1950

**Comment by rschu1ze** [comment]
All performance regressions are resolved. @yariks5s Thanks for going the extra mile!

**Comment by Slach** [comment]
@yariks5s will toDate, toDateTime, toDateTime64 also support non-constant time zones support?

**Comment by climber73** [comment]
Hey @yariks5s ! What about `toStartOf*` functions that have an optioal timezone argument?
Now the following:
```sql
toStartOfDay(datetime_column, string_column)
```
returns:
```
Received exception from server (version 23.12.4):
Code: 44. DB::Exception: Received from localhost:9000. DB::Exception: Illegal column string_column of time zone argument of function, must be a constant string: While processing toStartOfDay(datetime_column, string_column). (ILLEGAL_COLUMN)
```

**Comment by yariks5s** [comment]
>Hey @yariks5s ! What about toStartOf* functions that have an optioal timezone argument?
> Now the following:
> 
> toStartOfDay(datetime_column, string_column)
> returns:
> 
> Received exception from server (version 23.12.4):
> Code: 44. DB::Exception: Received from localhost:9000. DB::Exception: Illegal column string_column of time zone argument of function, must be a constant string: While processing toStartOfDay(datetime_column, string_column). (ILLEGAL_COLUMN)

@climber73, Hi! According to the [docs](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions#tostartofday), we don't have a timezone argument for `toStartOf..` functions at the moment. Especially for me, it is not clear what is the purpose of it and what we need to have as a result, using different timezones as arguments. There is a possibility to specify a timezone in a datetime_column argument of this function, in my opinion, using a different timezone in the `toStartOf..` would be a bit misleading

**Comment by Avogar** [comment]
> So my question is why don't we allow non-constant for TZ args and return just a point in time that will be displayed according to the column TZ (if we use the result as values to be inserted there)

The problem with functions `toStartOf*` is that return type is `DateTime`, not `String`, so it stores only constant timezone because `DateTime` type doesn't support different timezones for different rows. We can apply non-constant timezone only if the result is `String` (so we can apply corresponding timezone before formatting).

**Comment by Avogar** [comment]
Maybe you can try to use `toString` with non-constant timezone to convert timestamp to the client timezone and after it apply `toStartOf*`:
```sql
create table test (id UInt64, ts DateTime, tz String) engine=Memory;
insert into test values (1, '2020-01-01 00:00:00', 'UTC'), (1, '2020-01-01 00:00:00', 'Europe/Amsterdam');
select ts, tz, toString(ts, tz), toStartOfDay(toDateTime(toString(ts, tz))) from test;
```

```
┌──────────────────ts─┬─tz───────────────┬─toString(ts, tz)────┬─toStartOfDay(toDateTime(toString(ts, tz)))─┐
│ 2020-01-01 23:00:00 │ UTC              │ 2020-01-01 23:00:00 │                        2020-01-01 00:00:00 │
│ 2020-01-01 23:00:00 │ Europe/Amsterdam │ 2020-01-02 00:00:00 │                        2020-01-02 00:00:00 │
└─────────────────────┴──────────────────┴─────────────────────┴────────────────────────────────────────────┘
```

**Comment by climber73** [comment]
Good catch, thank you! Are you going to apply this fix (for toString function) for other LTS versions, for example 22.3 LTE?

**Comment by Avogar** [comment]
> Are you going to apply this fix (for toString function) for other LTS versions, for example 22.3 LTE?

No, it was not a fix but improvement and we don't backport such PRs.

**Pull request merged**